pkg/controller/validation/check_snapshot.go

```diff
@@ -16,32 +16,48 @@ import (
 var log = logf.Log.WithName("check_snapshot")
 
 const (
-	snapshotCRDVersion = "v1alpha1"
-	snapshotCRDName    = "volumesnapshots.snapshot.storage.k8s.io"
+	alphaVersion           = "v1alpha1"
+	snapshotCRDName        = "volumesnapshots.snapshot.storage.k8s.io"
+	snapshotClassCRDName   = "volumesnapshotclasses.snapshot.storage.k8s.io"
+	snapshotContentCRDName = "volumesnapshotcontents.snapshot.storage.k8s.io"
 )
 
-// Returns an error if the v1alpha1 VolumeSnapshot CRD is installed in the cluster
+// Returns an error if the v1alpha1 CRD is installed in the cluster
 // and nil otherwise. This is used to prevent upgrading where clusters have manually
 // installed versions we don't support
+
 func CheckAlphaSnapshot(client client.Client) error {
-	snapshotCRD := &apiextv1beta1.CustomResourceDefinition{}
-	err := client.Get(context.TODO(), types.NamespacedName{Name: snapshotCRDName, Namespace: corev1.NamespaceAll}, snapshotCRD)
-	if err != nil {
-		if apierrors.IsNotFound(err) {
-			// If we can't find the VolumeSnapshot CRD, then we're safe to proceed
-			return nil
+	crdMap := map[string]string{"VolumeSnapshot": snapshotCRDName,
+		"VolumeSnapshotClass":   snapshotClassCRDName,
+		"VolumeSnapshotContent": snapshotContentCRDName}
+	msgString := ""
+	resource := ""
+	found := false
+	crd := &apiextv1beta1.CustomResourceDefinition{}
+	for k, v := range crdMap {
+		err := client.Get(context.TODO(), types.NamespacedName{Name: v, Namespace: corev1.NamespaceAll}, crd)
+		if err != nil {
+			if !apierrors.IsNotFound(err) {
+				errMsg := "Error attempting to obtain existing " + k + " CRD."
+				log.Error(err, errMsg)
+				return err
+			}
 		}
-		log.Error(err, "Error attempting to obtain existing VolumeSnapshot CRD")
-		return err
-	}
 
-	for _, version := range snapshotCRD.Spec.Versions {
-		if version.Name == snapshotCRDVersion {
-			err = apierrors.NewConflict(schema.GroupResource{Resource: "VolumeSnapshot"},
-				"v1alpha1 VolumeSnapshot installed.", nil)
-			log.Error(err, "Unable to update cluster as VolumeSnapshot v1alpha1 is detected. Remove this version to allow the upgrade to proceed.")
-			return err
+		for _, version := range crd.Spec.Versions {
+			if version.Name == alphaVersion {
+				found = true
+				resource = k
+				errString := "v1alpha1 " + k + " installed"
+				msgString = msgString + "Unable to update cluster as " + k + " " + alphaVersion + " is detected. Remove this version to allow the upgrade to proceed. "
```

Comment:
If all three CRDs are installed, then the message will be really long. Maybe:
"Unable to update cluster as v1alpha1 version of volumesnapshots.snapshot.storage.k8s.io, volumesnapshotcontents.snapshot.storage.k8s.io, volumesnapshotsclasses.snapshot.storage.k8s.io is detected. Remove this version to allow the upgrade to proceed"

i.e. `strings.Join(foundCRDs, ", ")`

Reply:
I've modified this message to build up a slice of the detected CRDs, which is used to create the final message. Since we're creating this slice, I also changed the condition from a boolean to check and see if the slice isn't empty.

Let me know your thoughts on this.